### PR TITLE
feat(new_metrics): implement the metric & its prototype

### PR DESCRIPTION
**Pull request by empiredan**
https://github.com/apache/incubator-pegasus/issues/928

### CODE REVIEW

**Review by acelyc111 — APPROVED** [review]
LGTM

File: include/dsn/utility/metrics.h [hunk 1]
```diff
@@ -117,4 +146,90 @@ class metric_registry : public utils::singleton<metric_registry>
...
+    {
+        const string_view entity_type;
+        const string_view name;
+        const metric_unit unit;
+        const string_view desc;
+    };
+
+    const string_view &entity_type() const { return _args.entity_type; }
+
+    const string_view &name() const { return _args.name; }
```
Comment (levy5307) [inline include/dsn/utility/metrics.h:178]: no need return reference for string_view

File: include/dsn/utility/metrics.h [hunk 2]
```diff
@@ -117,4 +146,90 @@ class metric_registry : public utils::singleton<metric_registry>
...
+// The ref count of a metric may becomes 1, which means the metric is only held by the entity:
+// After a period of configurable time, if the ref count is still 1, the metric will be dropped
+// in that it's considered to be useless. During the period when the metric is retained, once
+// the same one is instantiated again, it will not be removed; whether the metric is instantiated,
+// however, its lastest value is visible.
+class metric : public ref_counter
+{
+protected:
+    explicit metric(const metric_prototype *prototype);
+    virtual ~metric();
```
Comment (levy5307) [inline include/dsn/utility/metrics.h:227]: virtual ~metric() = default;

File: include/dsn/utility/casts.h [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+// pointer(reference). As a usual approach, RTTI (dynamic_cast<>) is not efficient.
+// Instead, we can perform a compile-time assertion check whether one is derived
+// from another; then, just use static_cast<> to do the conversion faster. RTTI is
+// also run in debug mode to do double-check.
+
+template <typename To, typename From>
+inline To down_cast(From *from)
+{
+    // Perform a compile-time assertion to check whether <From> class is derived from <To> class.
+    static_assert(std::is_base_of<typename std::remove_pointer<From>::type,
```
Comment (levy5307) [inline include/dsn/utility/casts.h:35]: This looks great to me :) :+1:

**Review by acelyc111 — APPROVED** [review]
LGTM

**Review by levy5307 — APPROVED** [review]

**Pull request merged**